### PR TITLE
Tabishop initial rest projects integration

### PR DESCRIPTION
**Pull request by tabishop**
This PR has the code to include the REST projects and stubbed API endpoints. These will later be updated with the ability to call the cohorting engine, but for now, they are just stubs. 

The cohort-engine-api-web project gets built into the .war file and the cohort-engine-api project contains all the model objects that will be used by the rest api. I created separate projects in case we wanted to use the model objects within other projects, having this separated from the REST endpoint code will allow us to avoid having circular compile dependencies.

There were a lot of changes to the cohort-parent POM files. I used the template projects from the common services team for these and may need to contact that team with questions if you see things that you want changed.

Also please note the updated instructions for building while referencing the rest-service-framework projects (docs\dev-guide\getting-started.md) if you want to pull this branch and build it locally.

### CODE REVIEW

File: cohort-engine-api-web/launchers/cohort-engine-api-web-VT.launch
Comment (jillrdoty) [inline cohort-engine-api-web/launchers/cohort-engine-api-web-VT.launch:1]: I generally prefer not to commit eclipse specific files, but I'm unfamiliar with the purpose of this one so not sure if it could be auto-generated by anyone who wishes to launch in eclipse?

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplication.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplication.java:46]: nitpick: looks like a mix of tabs and spaces in this file

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java:15]: Should this be removed as the variable is configured on line 19?

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java:21]: Curious as to the purpose for these two parameters?

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java:44]: I know this isn't really doing anything yet, but is there a reason to comment out the logger?

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestStatusHandler.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestStatusHandler.java:31]: Should the state of the service be set to the value passed in instead of OK always?

Also I'm intrigued by an endpoint that allows an external caller to update the state of the service - can you share a real-life use case for this?

File: cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplicationTest.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplicationTest.java:1]: Copyright needed

File: cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstantsTest.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstantsTest.java:1]: Copyright needed

File: cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandlerTest.java
Comment (jillrdoty) [inline cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandlerTest.java:1]: Correct copyright needed

File: cohort-parent/pom.xml
Comment (jillrdoty) [inline cohort-parent/pom.xml:72]: Curious what we're doing with setting skipTests to true?

File: cohort-parent/pom.xml
Comment (jillrdoty) [inline cohort-parent/pom.xml:75]: Lines 75 and 76 are duplicates of 34 and 35?

File: cohort-parent/pom.xml
Comment (jillrdoty) [inline cohort-parent/pom.xml:293]: nitpick: cleanup commented dependency (I assume duplicated)

File: cohort-parent/pom.xml
Comment (jillrdoty) [inline cohort-parent/pom.xml:496]: I appreciate the plugin comments explaining their purpose.

File: docs/dev-guide/getting-started.md
Comment (jillrdoty) [inline docs/dev-guide/getting-started.md:33]: Should we be recommending write and delete packages for personal tokens?

File: cohort-engine-api-web/launchers/cohort-engine-api-web-VT.launch
Comment (tabishop) [reply cohort-engine-api-web/launchers/cohort-engine-api-web-VT.launch:1]: This file is used to launch a junit test suite from eclipse with 1 click which is very handy. It can't fully be autogenerated because it would require knowledge of the VM arguments which would be difficult to dig up for someone who isn't very familiar with the code. I would prefer to keep it in, unless it is causing a problem on other IDEs because it is something I use all the time.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplication.java
Comment (tabishop) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplication.java:46]: Good catch, will cleanup.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java
Comment (tabishop) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java:15]: It could be, but it is part of the comment illustrating of how to use include multiple packages using a comma separated list and I think contains useful information.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java
Comment (tabishop) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstants.java:21]: The rest service framework provides a handy means for having REST API calls that can be dark launched via env vars. The mechanism includes hiding them from the OpenAPI/Swagger interface when they are disabled. They would be used on the REST API definition and in the env vars to enable/disable certain API calls. These were 2 leftover examples of that from the sample application which I will cleanup.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java
Comment (tabishop) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java:44]: Just to remove an unused variable warning in eclipse. I'm happy to uncomment it as it will be used when these REST apis are fleshed out.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestStatusHandler.java
Comment (tabishop) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestStatusHandler.java:31]: The actual liveness, readiness, and healthcheck endpoints are provided by the superclass ServiceStatusHandler, which is part of the rest service framework projects. The superclass endpoint methods call this method to allow the subclass to be in the implementing project and this is where the application specific logic for determining the state of the application would go. The endpoints themselves don't allow and external caller to pass in the service state. In the future we may want to add logic here to check if any configured FHIR servers, databases, or any other external services that are needed for successful operation of our service are up and can be reached.

File: cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplicationTest.java
Comment (tabishop) [reply cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestApplicationTest.java:1]: Good catch, will add.

File: cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstantsTest.java
Comment (tabishop) [reply cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestConstantsTest.java:1]: Good catch, will add.

File: cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandlerTest.java
Comment (tabishop) [reply cohort-engine-api-web/src/test/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandlerTest.java:1]: Good catch, will add.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java
Comment (tabishop) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java:44]: I was wrong. Having it uncommented causes a PMD violation and fails the build <violation beginline="47" endline="47" begincolumn="37" endcolumn="42" rule="UnusedPrivateField" ruleset="Best Practices" package="com.ibm.cohort.engine.api.service" class="CohortEngineRestHandler" variable="logger" externalInfoUrl="https://pmd.github.io/pmd-6.21.0/pmd_rules_java_bestpractices.html#unusedprivatefield" priority="3">
Avoid unused private fields such as 'logger'.
</violation>

File: cohort-parent/pom.xml
Comment (tabishop) [reply cohort-parent/pom.xml:72]: This was pulled over from the example project definition and I think it was an attempt to allow skipping tests on a per-project basis (ala https://maven.apache.org/surefire/maven-surefire-plugin/examples/skipping-tests.html). I don't think this is really needed for our projects at this time (we can still use the maven.test.skip property if we want in our dev envs.). I am going to remove this property and it's references until such time as we determine that it is actually needed.

File: cohort-parent/pom.xml
Comment (tabishop) [reply cohort-parent/pom.xml:75]: Good catch, will cleanup.

File: cohort-parent/pom.xml
Comment (tabishop) [reply cohort-parent/pom.xml:293]: Good catch, will cleanup.

File: cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java
Comment (jillrdoty) [reply cohort-engine-api-web/src/main/java/com/ibm/cohort/engine/api/service/CohortEngineRestHandler.java:44]: Interesting, thanks

File: docs/dev-guide/getting-started.md
Comment (tabishop) [reply docs/dev-guide/getting-started.md:33]: Good point, confirmed it works with just read:packages and will remove the others

**Review by jillrdoty — APPROVED** [review]
Approving, but would ideally like another set of eyes on things.

File: docs/dev-guide/getting-started.md
Comment (csandersdev) [inline docs/dev-guide/getting-started.md:33]: It appears that the github actions are working for your branch, but it isn't clear to me how without some additional work to enable these settings. Have you thought through that interaction?

File: cohort-engine-api-web/src/main/resources/logback.xml
Comment (csandersdev) [inline cohort-engine-api-web/src/main/resources/logback.xml:17]: This is mostly just a comment. I jump through hoops in cohort-parent/pom.xml to remove transitive dependencies on logback from upstream OSS components because I don't believe that consumers of our library should be bound to a particular logging implementation and SLF4J likes to complain when you have more than one binding in your classpath. 

That said, I have no problem with the REST API layer being dependent on logback since it is functionally an app and nothing will link to it. I just want to make sure the logback dependency doesn't leak back into the library level code. I don't see that happening. Logback is never referenced in our pom's, so we must just be getting it transitively from the framework.

File: cohort-engine-api-web/src/test/resources/deploymentEnv.prop
Comment (csandersdev) [inline cohort-engine-api-web/src/test/resources/deploymentEnv.prop:7]: Hard-coded port? Is this specific to your local Liberty setup or will it be the same for every consumer of the code?

**Review by csandersdev — APPROVED** [review]
I left a few minor questions/comments, but I don't see anything that should block merge. We will probably want to rereview the API methods that we are exposing at some point, but I am fine as it is for now.

**Review by csandersdev — COMMENTED** [review]
I'm getting a warning for "overridding managed version" in reports/pom.xml for the Maven deploy plugin. You can fix by just removing the version. It is set in cohort-parent/pom.xml now.

File: cohort-engine-api-web/src/main/resources/logback.xml
Comment (tabishop) [reply cohort-engine-api-web/src/main/resources/logback.xml:17]: The logback dependency is something provided by the rest service framework and I copied it in based on how the example application used it. It is new since the last time I used the framework and I will not claim to be an expert in it. So your comments are welcome, we should keep an eye out to see if it causes any problems once we start calling the cohort engine jar from the rest api, as I am currently packaging the cohort jar in the .war, but not yet calling it.

File: cohort-engine-api-web/src/test/resources/deploymentEnv.prop
Comment (tabishop) [reply cohort-engine-api-web/src/test/resources/deploymentEnv.prop:7]: I will have to dig into this a bit more. It does get packaged with the tests jar that is intended to be used during the IBM Cloud toolchain deployment process, but unsure if it is actually used. It was pulled over from the example application.

Comment (tabishop) [reply cohort-engine-api-web/src/test/resources/deploymentEnv.prop]: Talked with Tom P. and determined that this file is no longer used so I'm removing it.

File: docs/dev-guide/getting-started.md
Comment (tabishop) [reply docs/dev-guide/getting-started.md:33]: Using GitHub Actions uses a GITHUB_TOKEN which is created automatically by GitHub. I added a reference to the build.yml to make sure it can use it and authenticate.

**Pull request merged**
